addons/account/models/account_move.py

```diff
@@ -1,349 +1,1869 @@
+
+        For example, tax lines depends of base lines (lines having tax_ids set). This is also the case of cash rounding
+        lines that depend of base lines or tax lines depending the cash rounding strategy. When a payment term is set,
+        this method will auto-balance the move with payment term lines.
+
+        :param recompute_all_taxes: Force the computation of taxes. If set to False, the computation will be done
+                                    or not depending of the field 'recompute_tax_line' in lines.
+        '''
+        self.ensure_one()
+
+        lines_map = {
+            'base_lines': self.env['account.move.line'],
+            'tax_lines': self.env['account.move.line'],
+            'terms_lines': self.env['account.move.line'],
+            'rounding_lines': self.env['account.move.line'],
+        }
+
+        # Dispatch lines and pre-compute some aggregated values like taxes.
+        for line in self.line_ids:
+            if line.recompute_tax_line:
+                recompute_all_taxes = True
+                line.recompute_tax_line = False
+                # Ensure the tags are correctly set.
+                if self.type in ('out_refund', 'in_refund'):
+                    tax_repartition_field = 'refund_repartition_line_ids'
+                else:
+                    tax_repartition_field = 'invoice_repartition_line_ids'
+                line.tag_ids = line.tax_ids\
+                    .mapped(tax_repartition_field)\
+                    .filtered(lambda line: line.repartition_type == 'base')\
+                    .mapped('tag_ids')
+
+            if self._is_invoice() and line.is_rounding_line:
+                lines_map['rounding_lines'] += line
+            elif line.tax_repartition_line_id:
+                lines_map['tax_lines'] += line
+            elif self._is_invoice() and not line.exclude_from_invoice_tab:
+                lines_map['base_lines'] += line
+            elif self._is_invoice() and line.account_id.user_type_id.type in ('receivable', 'payable'):
+                lines_map['terms_lines'] += line
+            else:
+                lines_map['base_lines'] += line
+
+        # Compute taxes.
+        if recompute_all_taxes:
+            self._recompute_tax_lines(lines_map)
+
+        if self._is_invoice():
+            # Compute totals.
+            # /!\ This computation is not done into the first loop as we need to take into account the newly updated
+            # tax lines.
+            totals_map = {
+                'total_balance': 0.0,
+                'total_amount_currency': 0.0,
+            }
+            for line in lines_map['base_lines']:
+                totals_map['total_balance'] += line.balance
+                totals_map['total_amount_currency'] += line.amount_currency
+            for line in lines_map['tax_lines']:
+                totals_map['total_balance'] += line.balance
+                totals_map['total_amount_currency'] += line.amount_currency
```

Comment:
Why not 
```python
for line in lines_map['base_lines'] + lines_map['tax_lines']:
    totals_map['total_balance'] += line.balance
    totals_map['total_amount_currency'] += line.amount_currency

Reply:
In term of performance, the current way is better as it doesn't create a new list